DSCResources/MSFT_xSQLAOGroupEnsure/MSFT_xSQLAOGroupEnsure.psm1

```diff
@@ -0,0 +1,351 @@
+        $SetupCredential
+    )
+
+   $null = [System.Reflection.Assembly]::LoadWithPartialName('Microsoft.SqlServer.Smo')
+   $null = [System.Reflection.Assembly]::LoadWithPartialName("Microsoft.SqlServer.SmoExtended")
+   
+   if(!$SQL)
+    {
+        $SQL = Connect-SQL -SQLServer $SQLServer -SQLInstanceName $SQLInstanceName -SetupCredential $SetupCredential
+    }
+
+   if (($AvailabilityGroupNameIP -and !$AvailabilityGroupSubMask) -or (!$AvailabilityGroupNameIP -and $AvailabilityGroupSubMask))
+    {
+        Throw "AvailabilityGroupNameIP and AvailabilityGroupSubMask must both be passed for Static IP assignment."
+        Exit
+    }
+
+   Switch ($Ensure)
+   {
+       "Present"
+       {
+           Grant-ServerPerms -SQLServer $SQLServer -SQLInstanceName $SQLInstanceName -AuthorizedUser "NT AUTHORITY\SYSTEM" -SetupCredential $SetupCredential
+           New-ListenerADObject -AvailabilityGroupNameListener $AvailabilityGroupNameListener -SetupCredential $SetupCredential
+           
+           $FailoverCondition = 3
+           $HealthCheckTimeout = 30000
+           $ConnectionModeInPrimary ="AllowAllConnections"    
+           $ConnectionModeInSecondaryRole = switch ($ReadableSecondary)
+                               {
+                                   'None' {"AllowNoConnections"}
+                                   'ReadOnly' {"AllowAllConnections"}
+                                   'ReadIntent'{"AllowReadIntentConnectionsOnly"}
+                                   Default {"AllowAllConnections"}
+                               } 
+
+           #Get Servers participating in the cluster
+           #First two nodes will account for Syncronous Automatic Failover, Any additional will be Asyncronous
+           try
+           {
+                $nodes= Get-ClusterNode -cluster $sql.ClusterName -Verbose:$false |  select -ExpandProperty name
+                $syncNodes = $nodes | Select-Object -First 2
+                $asyncNodes = $nodes | Select-Object -Skip 2
+                $availabilityGroup = New-Object -typename Microsoft.SqlServer.Management.Smo.AvailabilityGroup -ArgumentList $SQL, $AvailabilityGroupName
+                $availabilityGroup.AutomatedBackupPreference="Primary"
+                $availabilityGroup.FailureConditionLevel = $FailoverCondition
+                $availabilityGroup.HealthCheckTimeout =$HealthCheckTimeout
+           }
+           Catch
+           {
+                Throw "Failed to connect to Cluster Nodes from $sql.ClusterName"
+                Exit
+           }
+
+           #Loop through Sync nodes Create Replica Object Assign properties and add it to AvailabilityGroup
+           foreach ($node in $syncNodes)
+           { 
+               Try
+               {
+                    $Replica = New-Object -typename Microsoft.SqlServer.Management.Smo.AvailabilityReplica -ArgumentList $availabilityGroup, $node
+                    $Replica.EndpointUrl = "TCP://$($node):5022"
```

Comment:
5022 is used in many places, can you make it a const also describe what this magic number is for?